### DISCUSSION TITLE
Potential development of features for servers?

### DISCUSSION THREAD

**Discussion by wizardofgcc** [jeremyshannon/Exile#33] [Ideas]
So just today a server running Exile was opened: https://forum.minetest.net/viewtopic.php?f=10&t=27331&p=400884#p400884

So might this be a reason to add some more features oriented at servers? Having different skins would definetely be a big one, and that will probably work well with Runs' player API. Some other things that might be in the core game for servers:
- Locks - probably the simplest. Maybe should only work with wood/iron doors to motivate people to use them?
- Clothing customization - dyes. Even with the random skins this would be useful, since when people have the cana fiber clothing they will probably look the same. I'm not sure about this one though, maybe will be better for a separate mod
- Signs - allow writing something with letters in the world. Also not sure about this one, communicating with pictograms seems more fitting, since lore-wise all the characters probably don't speak the same language. Also maybe for a separate mod?

**Comment by jeremyshannon** [comment]
Different skins and dying clothing are things I've thought about.  Locks of some kind would be nice, too. I've been thinking of expanding the pictograms as well. (There was an old video game called Captain Blood where you communicated with aliens by stringing together different pictograms, it was pretty neat)

 I figure skins should be the next big thing I work on, I did a bit of work looking into runs' api, but I may just port a couple things over instead of switching entirely. I figure random skins could have multiple skin tones, eye/hair color, maybe apply scars including ritual scarification on a few. Could have non-human skin tones as well. The Ancients spread far and wide, and the folks who are using their portals may come from all sorts of places. Different continent, different planet, different dimension, who knows?

**Comment by wizardofgcc** [comment]
Yea, the random skins were planned for a long time. Expanding the pictograms would be good, too. The comments in tech/grafitti.lua suggest two more colours: ochre (i guess from crushed ironstone) and black (probably from charcoal). Those could be added to separate pictograms of different themes (e. g. biomes, trail marks, etc), since can't just put everything into the lime white one. 
On locks I have a couple ideas:
The simpler, less realistic one - make locked chest/door and a key. Whoever places the locked chest can open it/break it/etc or make a key and bind it to give that privilege to someone else (can interact if key is in inventory).
More complex, more realistic - lock and key are made as one item, right click door/chest with that item to create locked chest and a matching key. Right click chest/door with key to open it. This has issues though, mainly copying keys.

**Comment by jeremyshannon** [comment]
Yeah, for locks, maybe you make a lockable chest, and when initially placed, it's "chest (unlocked)" until you make a key and use it. Copying keys would be doable by anyone who had iron to make a copy and possessed the key to be copied.  Players would be encouraged to have spare keys stored by a friend, or buried somewhere hard to find.  There might be something in there for a locksmith profession in the roleplaying ideas, but I'm not sure I'd want to require one to spawn before you could protect anything.

 I think antiquorium doors could have special locks that when activated, would only open for the player they're tuned to, and wouldn't need a key. Though obviously an antiquorium chisel would still be able to take them off their hinges. I wonder if I could make that only work from the inside? Making players use chisels to break through the walls of undercity locations that have their doors closed would be nasty, but I can't deny it has a certain appeal.

**Comment by jeremyshannon** [comment]
Late night, random thoughts on this, inspired by events on the LoC server.
A player was on, trying to skip straight to iron without even bothering to make clothing -- the static spawn location on LoC meant everytime he died he could just walk over and pick up where he left off. (He was using his bones as a temporarily locked storage container, which is pretty ingenious IMO -- you just pick up all your important stuff before you die.)

 This made me think about protection and spawn points. The main disincentive to play the game this way normally in singleplayer is that you always spawn somewhere new, and you have to find your way back to your base. On LoC, that's not a concern, so the main disincentive is that you can't protect your stuff from thieves and vandals. (Though there's been surprisingly little of that, he and some other players had a grindstone and other valuable things sitting out for a few days and nobody took them. I think my  protecting a few things near spawn means the usual tourists assume everything nearby is protected and they stop trying)

If protection is available to pre-iron players, static spawn means you can ignore the survival aspect of the game entirely. I was thinking of making it possible to bar a wattle door from inside with a stick, but i'm not sure about that now.

A better spawn system could help, perhaps allowing server owners to set up multiple spawn points, but unless there's a lot, you can always throw yourself down a pit and reroll until you get the one you were using. Perhaps spawn players close to the center of the map at first, then start dropping them further out based on something? We could maybe use beds, the ones in use are already recorded --  then players respawn at least X mapblocks away from any occupied bed.
 This makes it possible to build a town and defend it from intruding thieves; killing them will cause them to re-spawn far away. Especially after MT 5.5.0, when coordinates are no longer visible to players, finding your way back will be difficult.

**Comment by jeremyshannon** [comment]
More concrete thoughts on this, written down here so I don't forget.

A good multiplayer spawn system for Exile would mimic the native singleplayer experience but on a wider scale. By default, Minetest randomly drops newly spawned characters somewhere around the center of the map. This means you might be reasonably far away from where your base is, but not ludicrously so, making death a pain, but not a reason to give up on ever finding your old base.

 For multiplayer, I'd prefer to do this in a region-by-region way. If we divide the map up into 2048-by-2048 sectors, we get 256 of these "neighborhoods" in a world. The size is large enough that you can't easily just walk across to the next sector naked and freezing to die and respawn there to continue your journey, you'd need at least a bit of preparation and care. It's also large enough that when you die at your base, you can't just spin the camera around and say "oh, there it is." We might also require you to enter a ways past the edge of a sector before setting your spawn point there.

 Players would begin at the center of the map, or at an admin-designated sector, and would then be subject to the regional respawn system. Travelling to a player-built city in some corner of the world would allow them to respawn in the neighborhood around it when they die.

This could be combined with the above idea: players need to place at least a primitive_bed down indoors before it will change their respawn point to a specific neighborhood.
